Add parameter to forwards client calls to support APIs that cache results

m-nash opened the pull request:
# All SDK Contribution checklist:

This allows us to turn off DiagnosticScopeValidation for a single method on a client vs today having to turn on or off the entire client.

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

m-nash committed:
support testing APIs that cache results

m-nash committed:
add test case to validate true and false

annelo-msft reviewed (commented):
Curious: why do you want to turn off diagnostic scope validation for a single call?

christothes commented on sdk/core/Azure.Core.TestFramework/src/Instrumentation/DiagnosticScopeValidatingInterceptor.cs:
```diff
@@ -180,6 +181,10 @@ internal static async ValueTask<T> ValidateDiagnosticScope<T>(Func<ValueTask<(T
             {
                 // Remove subscribers before enumerating events.
                 diagnosticListener.Dispose();
+                bool? skipOverride = forwardAttribute is null
+                    ? default(bool?)
+                    : (bool?)forwardAttribute.GetType().GetProperty("SkipChecks").GetValue(forwardAttribute);
+                skipChecks |= skipOverride.HasValue && skipOverride.Value;
```
Since we know the type of the property is bool, should we cast as (bool) and simplify the need to check HasValue?

m-nash replied:
it can be null for the default constructor so if we call .Value or try to cast to bool we will get an exception

christothes replied:
Does it need to be nullable at all? The default value would be `false` and work the same way, I believe.

m-nash replied:
I think the challenge is this needs to get merged with the global client level skipChecks.  I think we need to differentiate from "not set" and "explicit false" that way we only override the global client level setting when you explicitly set it

m-nash replied:
I guess its not technically an override its an also.  Here is how i understand the expected behavior

(Class, Method)
(true, true) = skip checks
(true, false) = skip checks
(false, true) = skip checks
(false, false) = don't skip checks|

If we expand that to include bool? the results don't change since its not override the class `false` its only overriding the class `true`

christothes replied:
I may be missing something, but doesn't `skipChecks` evaluate to the same value regardless if the value is `skipOverride` is `null` vs `false`?

christothes replied:
```c#
bool skipChecks = false;
bool? skipOverride = null;
bool boolSkip = false;
bool nullResult = skipChecks |= skipOverride.HasValue && skipOverride.Value;
bool boolResult = skipChecks |= boolSkip;
Console.WriteLine(nullResult);  //False
Console.WriteLine(boolResult); //False
```

m-nash replied:
yeah, if it was a method level override we need the null.  I think I am agreeing with you now that it isn't an override its an OR against the class level value.

m-nash commented:
> Curious: why do you want to turn off diagnostic scope validation for a single call?

If the call caches the result sometimes the diagnostic scope will not be there.  In this case there is an internal API that a public API calls so we need ForwardsClientCalls, but the public API will cache the results so the diagnostic scope won't always be there

christothes commented:
> > Curious: why do you want to turn off diagnostic scope validation for a single call?
> 
> If the call caches the result sometimes the diagnostic scope will not be there. In this case there is an internal API that a public API calls so we need ForwardsClientCalls, but the public API will cache the results so the diagnostic scope won't always be there

We should document this as a comment on the `SkipChecks` property

m-nash commented on sdk/core/Azure.Core/tests/TestClients/TestResource.cs:
```diff
@@ -51,6 +51,45 @@ public virtual Task<TestLroOperation> GetLroAsync(bool exceptionOnWait = false,
+            return Response.FromValue(new TestResource(), new MockResponse(200));
+        }
+
+        [ForwardsClientCalls(false)]
```
They should be yes, but I wanted to explicitly test both cases to validate that.  The tests that execute False and Default assert the same result

m-nash committed:
add comment on new property
fix back compat issue if you are using an older client where the property didn't exist

m-nash committed:
update to bool since this is an OR against the class value not an override to it.

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/Instrumentation/DiagnosticScopeValidatingInterceptor.cs:
```diff
@@ -180,6 +181,9 @@ internal static async ValueTask<T> ValidateDiagnosticScope<T>(Func<ValueTask<(T
             {
                 // Remove subscribers before enumerating events.
                 diagnosticListener.Dispose();
+                var skipOverrideProperty = forwardAttribute is not null ? forwardAttribute.GetType().GetProperty("SkipChecks") : null;
```
Why not cast this to the attribute type and use the property directly rather than using reflection?

m-nash replied:
Since its internal shared code this won't work well given different projects can have different versions of the class.
